File: day9/src/main.rs
use std::{collections::HashSet, fmt::Debug};

fn combinations(vec: &[Tile]) -> Vec<(Tile, Tile)> {
    let mut output = vec![];
    for i in 0..(vec.len() - 1) {
        for j in (i + 1)..vec.len() {
            output.push((vec[i], vec[j]));
        }
    }
    output
}

fn area((a, b): (&Tile, &Tile)) -> u64 {
    (a.0.abs_diff(b.0) + 1) * (a.1.abs_diff(b.1) + 1)
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Hash)]
struct Tile(u64, u64);

impl Tile {
    fn new(x: u64, y: u64) -> Tile {
        Tile(x, y)
    }
}

#[derive(Debug, PartialEq, Eq, Hash)]
enum Axis {
    Vertical = 0,
    Horizontal = 1,
}

#[derive(Debug, PartialEq, Eq, Hash)]
struct Edge {
    along: u64, // Vertical edge has shared x value, horizontal shared y
    start: u64,
    end: u64,
}

struct Part2 {
    tiles: Vec<Tile>,
    vertical_edges: HashSet<Edge>,
    horizontal_edges: HashSet<Edge>,
}

impl Part2 {
    fn new(tiles: &[Tile]) -> Self {
        Part2 {
            tiles: tiles.to_vec(),
            vertical_edges: find_edges(tiles, Axis::Vertical),
            horizontal_edges: find_edges(tiles, Axis::Horizontal),
        }
    }

    fn area_contains_no_edges(&self, t1: &Tile, t2: &Tile) -> bool {
        let (min_x, max_x) = (t1.0.min(t2.0), t1.0.max(t2.0));
        let (min_y, max_y) = (t1.1.min(t2.1), t1.1.max(t2.1));

        if self.vertical_edges.iter().any(|edge| {
            min_x < edge.along && edge.along < max_x && edge.end > min_y && edge.start < max_y
        }) {
            return false;
        }
        if self.horizontal_edges.iter().any(|edge| {
            min_y < edge.along && edge.along < max_y && edge.end > min_x && edge.start < max_x
        }) {
            return false;
        }
        true
    }

    fn solve(&self) -> u64 {
        combinations(&self.tiles)
            .iter()
            .filter(|(t1, t2)| self.area_contains_no_edges(t1, t2))
            .map(|(t1, t2)| area((t1, t2)))
            .max()
            .unwrap()
    }
}

/// Returns all the edges that are along the given axis
fn find_edges(tiles: &[Tile], axis: Axis) -> HashSet<Edge> {
    use Axis::*;
    combinations(tiles)
        .iter()
        .filter(|(t1, t2)| match axis {
            Vertical => t1.0 == t2.0,
            Horizontal => t1.1 == t2.1,
        })
        .map(|(t1, t2)| {
            let (along, start, end) = match axis {
                Vertical => (t1.0, t1.1.min(t2.1), t1.1.max(t2.1)),
                Horizontal => (t1.1, t1.0.min(t2.0), t1.0.max(t2.0)),
            };
            Edge { along, start, end }
        })
        .collect::<HashSet<_>>()
}

fn part1(tiles: &[Tile]) -> u64 {
    combinations(tiles)
        .iter()
        .map(|(t1, t2)| area((t1, t2)))
        .max()
        .unwrap()
}

fn main() {
    let data = std::fs::read_to_string("puzzle.txt").unwrap();
    let tiles = data
        .lines()
        .map(|line| line.split(",").collect::<Vec<&str>>())
        .map(|v| Tile::new(v[0].parse::<u64>().unwrap(), v[1].parse::<u64>().unwrap()))
        .collect::<Vec<_>>();

    println!("{:?}", part1(&tiles));
    let part2 = Part2::new(&tiles);
    println!("{:?}", part2.solve());
}

File: day9/main.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <numeric>
#include <ranges>
#include <unordered_set>
#include <vector>

using namespace std;

struct Point {
    uint64_t x;
    uint64_t y;

    Point(uint64_t x, uint64_t y) : x(x), y(y) {}

    uint64_t area(Point const& point) const {
        return (abs(static_cast<int64_t>(y - point.y)) + 1) *
               (abs(static_cast<int64_t>(x - point.x)) + 1);
    }
};

vector<pair<Point, Point>> combinations(vector<Point> const& points) {
    vector<pair<Point, Point>> result;
    for (int i = 0; i < points.size() - 1; i++) {
        for (int j = i + 1; j < points.size(); j++) {
            result.emplace_back(points[i], points[j]);
        }
    }
    return result;
}

enum class Axis { Vertical, Horizontal };

struct Edge {
    uint64_t along;
    uint64_t start;
    uint64_t end;

    bool operator==(Edge const& other) const {
        return along == other.along && start == other.start && end == other.end;
    }
};

class MyHashFunction {
   public:
    size_t operator()(const Edge& e) const { return (e.along * 100) ^ (e.start * 10) ^ e.end; }
};

using MySet = unordered_set<Edge, MyHashFunction>;

MySet find_edges(vector<Point> const& points, Axis axis) {
    return combinations(points) | views::filter([&axis](pair<Point, Point> const& pair) {
               auto [p1, p2] = pair;
               if (axis == Axis::Vertical) {
                   return p1.x == p2.x;
               } else {
                   return p1.y == p2.y;
               }
           }) |
           views::transform([&axis](pair<Point, Point> const& pair) {
               auto [p1, p2] = pair;
               if (axis == Axis::Vertical) {
                   return Edge{p1.x, min(p1.y, p2.y), max(p1.y, p2.y)};
               } else {
                   return Edge{p1.y, min(p1.x, p2.x), max(p1.x, p2.x)};
               }
           }) |
           ranges::to<MySet>();
}

struct Part2 {
    vector<Point> points;
    MySet vertical_edges;
    MySet horizontal_edges;

    Part2(vector<Point> const& points)
        : points(std::move(points)),
          vertical_edges(find_edges(points, Axis::Vertical)),
          horizontal_edges(find_edges(points, Axis::Horizontal)) {}

    bool area_contains_no_edges(Point const& p1, Point const& p2) {
        auto [min_x, max_x] = minmax(p1.x, p2.x);
        auto [min_y, max_y] = minmax(p1.y, p2.y);
        for (auto const& edge : vertical_edges) {
            if (min_x < edge.along && edge.along < max_x && edge.end > min_y &&
                edge.start < max_y) {
                return false;
            }
        }
        for (auto const& edge : horizontal_edges) {
            if (min_y < edge.along && edge.along < max_y && edge.end > min_x &&
                edge.start < max_x) {
                return false;
            }
        }
        return true;
    }

    uint64_t solve() {
        auto areas = combinations(points) | views::filter([&](pair<Point, Point> const& pair) {
                         auto [p1, p2] = pair;
                         return area_contains_no_edges(p1, p2);
                     }) |
                     views::transform([&](pair<Point, Point> const& points) {
                         auto [p1, p2] = points;
                         return p1.area(p2);
                     }) |
                     ranges::to<vector<uint64_t>>();
        return ranges::fold_left(
            areas, 0ull, [](uint64_t max_area, uint64_t area) { return max(max_area, area); });
    }
};

uint64_t part1(vector<Point> const& points) {
    auto combos = combinations(points);
    uint64_t max_area = transform_reduce(
        combos.begin(), combos.end(), 0ull,
        [](uint64_t max_area, uint64_t area) { return max(max_area, area); },
        [&](pair<Point, Point> const& combo) {
            auto const& [p1, p2] = combo;
            return p1.area(p2);
        });

    return max_area;
}

int main() {
    ifstream file("puzzle.txt");

    vector<Point> points;
    string line;
    while (getline(file, line)) {
        size_t pos = line.find(",");
        int x = stoi(line.substr(0, pos));
        int y = stoi(line.substr(pos + 1));
        points.push_back(Point(x, y));
    }

    println("{}", part1(points));
    Part2 part2(points);
    println("{}", part2.solve());
    return 0;
}
